### PR TITLE
Check _WINE_HWND_STYLE flags for POPUP

### LINKED ISSUE #392: Forza Horizon 5 starts "embedded" into steam
### Operating System

Arch Linux

### Wayland Compositor

Niri

### Describe the issue.

Starting with [v0.8.1](https://github.com/Supreeeme/xwayland-satellite/releases/tag/v0.8.1) Forza Horizon 5 starts "embedded" into Steam. It is not a floating window as I cannot drag it around or resize it.

Image

I also tried this with the latest commit (10f985b), and the issue persists.

### Steps to reproduce

- Launch Forza Horizon 5 on Steam with xwayland-satellite v0.8.1

### Please confirm you have reproduced this on the latest commit.

- [x] I have reproduced this on the latest commit

### PR DESCRIPTION
**Pull request by GoranKovac**
Steam games provide this Atom. Since `_MOTIF_WM_HINTS` can fail on them we check _WINE_HWND_STYLE` flags: https://learn.microsoft.com/en-us/windows/win32/winmsg/window-styles
https://learn.microsoft.com/en-us/windows/win32/winmsg/extended-window-styles

We determine by `WS_POPUP, WS_CAPTION, WS_THICKFRAME` + `WS_EX_APPWINDOW, WS_EX_TOOLWINDOW` is the wine window a popup.

Hopefully fixes https://github.com/Supreeeme/xwayland-satellite/issues/392
Fixes also: https://github.com/Supreeeme/xwayland-satellite/issues/392#issuecomment-4006218161

Setting as draft since it needs testing + more work (breaks other steam apps).

EDIT: Latest commit tested across these games/apps no regressions seen:
image

NOTE: This fix affects only games run thru steam/proton

### CODE REVIEW

**Comment by GoranKovac** [comment]
Needs more work, since its breaking some steam apps

**Comment by GoranKovac** [comment]
Latest commit works without any regression on various steam games/apps with no regressions

**Comment by GoranKovac** [comment]
Both issues have been confirmed fixed and I did not encountered any regressions

**Comment by Supreeeme** [comment]
Honestly, I'd really rather not check for these flags, because I'm pretty sure they're exclusive to Proton, and I just feel a bit icky about it. If there's no other option it is what it is though.
To me though, what stands out about the windows in that issue is that they all have the `motif::Functions::Close` and `motif::Functions::Move` flag. Maybe it's time to reconsider whether either of these flags truly indicate a popup? Adding a check for `motif::Functions::Close` doesn't break any tests for example, but I of course have not tested it extensively.

**Comment by GoranKovac** [comment]
> Honestly, I'd really rather not check for these flags, because I'm pretty sure they're exclusive to Proton, and I just feel a bit icky about it. If there's no other option it is what it is though. To me though, what stands out about the windows in that issue is that they all have the `motif::Functions::Close` and `motif::Functions::Move` flag. Maybe it's time to reconsider whether either of these flags truly indicate a popup? Adding a check for `motif::Functions::Close` doesn't break any tests for example, but I of course have not tested it extensively.

Sure will test (need to download few games).

**Comment by GoranKovac** [comment]
Close and move flag appear in many popups.... so we would need another solution.

**Comment by awsms** [comment]
> Honestly, I'd really rather not check for these flags, because I'm pretty sure they're exclusive to Proton, and I just feel a bit icky about it. If there's no other option it is what it is though.

I mean, why not as long as they're reliable? That's a Proton-only thing so far, that's mainly used by gamescope.
There was an attempt to get em merged in Wine as well but still nothing. https://gitlab.winehq.org/wine/wine/-/merge_requests/2902

**Comment by ArijanJ** [comment]
Been happening for around a week for me too, would appreciate if this was merged so we can play games :)

**Comment by ArijanJ** [comment]
I just tried this PR and it doesn't fix it actually, game still renders over the steam window and tiles weird on niri (and doesn't take keyboard input?):
```xprop
_NET_WM_BYPASS_COMPOSITOR(CARDINAL) = 0
WM_STATE(WM_STATE):
		window state: Normal
		icon window: 0x0
_NET_WM_ICON(CARDINAL) = 	Icon (32 x 32):
	[...]

_WINE_ALLOW_FLIP(CARDINAL) = 0
_NET_WM_NAME(UTF8_STRING) = "KovaaK’s  "
WM_ICON_NAME(COMPOUND_TEXT) = "KovaaK’s  "
WM_NAME(COMPOUND_TEXT) = "KovaaK’s  "
_WINE_HWND_EXSTYLE(CARDINAL) = 262144
_WINE_HWND_STYLE(CARDINAL) = 2483027968
WM_HINTS(WM_HINTS):
		Client accepts input or input focus: False
		Initial state is Normal State.
		bitmap id # to use for icon: 0x5e00051
		bitmap id # of mask for icon: 0x5e00053
		window id # of group leader: 0x6000003
_NET_WM_WINDOW_TYPE(ATOM) = _NET_WM_WINDOW_TYPE_NORMAL
_MOTIF_WM_HINTS(_MOTIF_WM_HINTS) = 0x3, 0x26, 0x0, 0x0, 0x0
WM_NORMAL_HINTS(WM_SIZE_HINTS):
		program specified location: 0, 0
		window gravity: Static
XdndAware(ATOM) = BITMAP
_NET_WM_PID(CARDINAL) = 73278
WM_LOCALE_NAME(STRING) = "en_US.UTF-8"
WM_CLIENT_MACHINE(STRING) = "arch"
WM_CLASS(STRING) = "steam_app_824270", "steam_app_824270"
WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW, _NET_WM_PING, WM_TAKE_FOCUS
_WINE_HWND(CARDINAL) = 65756
```

**Comment by GoranKovac** [comment]
Sorry work got in the way.... will take a look. There will be more edge cases but at some point Proton will enable native wayland support so this workaround will not be needed.

**Comment by GoranKovac** [comment]
@ArijanJ I have `kovaak` in my library so I can test but that game for some reason is  running weird (it cannot go in fullscreen no matter what). That issue is not related to this PR or has anything to do with whole `guess_popup` function (even when its bypassed so everything spawns as toplevel `kovaak`  spawns weirdly. So issue is somewhere else in xwsatellite

**Comment by ArijanJ** [comment]
Thanks for testing @GoranKovac, if it helps I can also reproduce the same issue with Mafia 3 with this PR:
```
⋊> ~ xprop
WM_STATE(WM_STATE):
		window state: Normal
		icon window: 0x0
_NET_WM_ICON(CARDINAL) = 	Icon (32 x 32):
[...]
_MOTIF_WM_HINTS(_MOTIF_WM_HINTS) = 0x3, 0x2c, 0x3a, 0x0, 0x0
_WINE_ALLOW_FLIP(CARDINAL) = 0
_NET_WM_NAME(UTF8_STRING) = "Mafia III: Definitive Edition"
WM_ICON_NAME(STRING) = "Mafia III: Definitive Edition"
WM_NAME(STRING) = "Mafia III: Definitive Edition"
_WINE_HWND_EXSTYLE(CARDINAL) = 256
_WINE_HWND_STYLE(CARDINAL) = 382337024
WM_HINTS(WM_HINTS):
		Client accepts input or input focus: False
		Initial state is Normal State.
		bitmap id # to use for icon: 0x4a00021
		bitmap id # of mask for icon: 0x4a00023
		window id # of group leader: 0x4c00003
_NET_WM_WINDOW_TYPE(ATOM) = _NET_WM_WINDOW_TYPE_NORMAL
WM_NORMAL_HINTS(WM_SIZE_HINTS):
		program specified location: 8, 29
		program specified minimum size: 1600 by 900
		program specified maximum size: 1600 by 900
		window gravity: Static
XdndAware(ATOM) = BITMAP
_NET_WM_PID(CARDINAL) = 108253
WM_LOCALE_NAME(STRING) = "en_US.UTF-8"
WM_CLIENT_MACHINE(STRING) = "arch"
WM_CLASS(STRING) = "steam_app_360430", "steam_app_360430"
WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW, _NET_WM_PING, WM_TAKE_FOCUS
STEAM_GAME(CARDINAL) = 360430
```
image

**Comment by awsms** [comment]
Codex was able to write a patch that successfully separates Proton windows from the others (to apply at the top of all the commits of this PR):
```
diff --git a/src/xstate/mod.rs b/src/xstate/mod.rs
index 48092fd..1d7ad83 100644
--- a/src/xstate/mod.rs
+++ b/src/xstate/mod.rs
@@ -636,6 +636,7 @@ impl XState {
         let size_hints = self.get_wm_size_hints(window);
         let motif_wm_hints = self.get_motif_wm_hints(window);
         let wine_hwnd_style = self.get_wine_hwnd_style(window);
+        let wine_hwnd_exstyle = self.get_wine_hwnd_exstyle(window);
         let wm_hints = self.get_wm_hints(window);
         let mut title = name.resolve()?;
         if title.is_none() {
@@ -657,6 +658,11 @@ impl XState {
             server_state.set_win_decorations(window, decorations);
         }
         let wine_style = wine_hwnd_style.resolve()?;
+        let wine_exstyle = wine_hwnd_exstyle.resolve()?;
+        let wine_hwnd = WineHwndStyle {
+            style: wine_style.and_then(|style| style.style),
+            ex_style: wine_exstyle.and_then(|style| style.ex_style),
+        };
         let transient_for = self
             .property_cookie_wrapper(
                 window,
@@ -673,7 +679,7 @@ impl XState {
             motif_hints,
             wmhints,
             transient_for.is_some(),
-            wine_style,
+            wine_hwnd,
         )?;
         server_state.set_popup(window, is_popup);
         if let Some(parent) = transient_for.and_then(|t| (!is_popup).then_some(t)) {
@@ -704,15 +710,15 @@ impl XState {
         motif_hints: Option<motif::Hints>,
         wm_hints: Option<WmHints>,
         has_transient_for: bool,
-        wine_style: Option<WineHwndStyle>,
+        wine_style: WineHwndStyle,
     ) -> XResult<bool> {
         let mut motif_popup = false;
         let mut wmhint_popup = false;
         let mut has_skip_taskbar = None;
         let mut is_wine_toplevel = false;
         // Used for steam/proton games, motif hints are not enough here to determine popup vs toplevel
-        if let Some(style) = wine_style {
-            is_wine_toplevel = !style.is_popup();
+        if wine_style.has_any_hint() {
+            is_wine_toplevel = !wine_style.is_popup();
         }
 
         let attrs = self
@@ -983,6 +989,25 @@ impl XState {
         }
     }
 
+    fn get_wine_hwnd_exstyle(
+        &self,
+        window: x::Window,
+    ) -> PropertyCookieWrapper<'_, impl PropertyResolver<Output = WineHwndStyle>> {
+        let cookie =
+            self.get_property_cookie(window, self.atoms.wine_hwnd_exstyle, x::ATOM_CARDINAL, 1);
+
+        let resolver = |reply: x::GetPropertyReply| {
+            let data: &[u32] = reply.value();
+            WineHwndStyle::from_ex_style(data)
+        };
+
+        PropertyCookieWrapper {
+            connection: &self.connection,
+            cookie,
+            resolver,
+        }
+    }
+
     fn get_pid(&self, window: x::Window) -> Option<u32> {
         let Some(pid) = self
             .connection
@@ -1089,6 +1114,7 @@ xcb::atoms_struct! {
         supported => b"_NET_SUPPORTED" only_if_exists = false,
         motif_wm_hints => b"_MOTIF_WM_HINTS" only_if_exists = false,
         wine_hwnd_style => b"_WINE_HWND_STYLE" only_if_exists = false,
+        wine_hwnd_exstyle => b"_WINE_HWND_EXSTYLE" only_if_exists = false,
         utf8_string => b"UTF8_STRING" only_if_exists = false,
         clipboard => b"CLIPBOARD" only_if_exists = false,
         clipboard_targets => b"_clipboard_targets" only_if_exists = false,
@@ -1200,27 +1226,32 @@ impl From<&[u32]> for WineHwndStyle {
 }
 
 impl WineHwndStyle {
+    fn from_ex_style(value: &[u32]) -> Self {
+        let mut ret = Self::default();
+
+        if let Some(&raw) = value.first() {
+            ret.ex_style = Some(Window32StyleEx::from_bits_truncate(raw));
+        }
+
+        ret
+    }
+
+    fn has_any_hint(&self) -> bool {
+        self.style.is_some() || self.ex_style.is_some()
+    }
+
     pub fn is_popup(&self) -> bool {
         if let Some(ex) = self.ex_style {
             if ex.contains(Window32StyleEx::WS_EX_APPWINDOW) {
                 return false;
             }
-            // Toolwindows are popups
             if ex.contains(Window32StyleEx::WS_EX_TOOLWINDOW) {
                 return true;
             }
         }
-
-        match self.style {
-            Some(s) => {
-                if !s.contains(Window32Style::WS_POPUP) {
-                    return false;
-                }
-                // caption or frame it toplevel
-                !s.intersects(Window32Style::WS_CAPTION | Window32Style::WS_THICKFRAME)
-            }
-            None => false,
-        }
+        // Proton frequently marks real toplevels as WS_POPUP for borderless/fullscreen windows.
+        // Style alone is too noisy to distinguish a popup from a standalone toplevel.
+        false
     }
 }
 ```
